ert_shared/tracker/evaluator.py

```diff
@@ -46,64 +46,67 @@ def __init__(self, model, ee_monitor_connection_details, states):
-                                drainer_logger.debug("connecting to new monitor...")
-                                monitor = create_ee_monitor(
-                                    self._monitor_host, self._monitor_port
-                                )
-                                wait_for_ws(monitor.get_base_uri(), max_retries=2)
-                                drainer_logger.debug("connected")
-                                break
-                            except ConnectionRefusedError as e:
-                                drainer_logger.debug(f"connection refused: {e}")
-                                pass
+
+                drainer_logger.debug("connecting to new monitor...")
+                with create_ee_monitor(
+                    self._monitor_host, self._monitor_port
+                ) as monitor:
+                    drainer_logger.debug("connected")
+                    for event in monitor.track():
+                        if event["type"] == ids.EVTYPE_EE_SNAPSHOT:
+                            iter_ = event.data["metadata"]["iter"]
+                            with self._state_mutex:
+                                self._realization_progress[
+                                    iter_
+                                ] = self._snapshot_to_realization_progress(event.data)
+                                self._work_queue.put(None)
+                                if event.data.get("status") == _EVTYPE_SNAPSHOT_STOPPED:
+                                    drainer_logger.debug(
+                                        "observed evaluation stopped event, signal done"
+                                    )
+                                    monitor.signal_done()
+                        elif event["type"] == ids.EVTYPE_EE_SNAPSHOT_UPDATE:
+                            with self._state_mutex:
+                                self._updates.append(event.data)
+                                self._work_queue.put(None)
+                                if (
+                                    event.data.get("status")
+                                    == _EVTYPE_SNAPSHOT_CANCELLED
+                                ):
+                                    drainer_logger.debug(
+                                        "observed evaluation cancelled event, return"
+                                    )
+                                    return
+                                if event.data.get("status") == _EVTYPE_SNAPSHOT_STOPPED:
+                                    drainer_logger.debug(
+                                        "observed evaluation stopped event, signal done"
+                                    )
+                                    monitor.signal_done()
+                        elif event["type"] == ids.EVTYPE_EE_TERMINATED:
+                            drainer_logger.debug("got terminator event")
+
+                # This sleep needs to be there. Refer to issue #1250: `Authority
```

Comment:
Some more details:
When an evaluation is finished, the next state is one of two:

1. `model` will finish, which means the experiment is done.
2. `model` will not finish, which _most likely_ means there's another evaluation coming down the pipe (iteration n+1)

You can't upfront know how many iterations the experiment needs (the data is probably somewhere in ERT/libres, but you someone smarter than me will have to find it), so we can only guess what it is going to be. Sleeping 5 seconds will usually let the model finish (but not always, so this sleep is incredibly bad).